apps/total-typescript/src/server/routers/pricing.ts

```diff
@@ -0,0 +1,143 @@
+import {createRouter} from 'server/createRouter'
+import {z} from 'zod'
+import {getSdk, Purchase} from '@skillrecordings/database'
+import {
+  formatPricesForProduct,
+  getActiveMerchantCoupon,
+} from '@skillrecordings/commerce-server'
+import {find} from 'lodash'
+import type {FormattedPrice} from '@skillrecordings/commerce-server/dist/@types'
+
+const merchantCouponSchema = z.object({
+  id: z.string(),
+  type: z.string(),
+})
+
+const PurchaseSchema: z.ZodType<Purchase> = z.any()
+
+const PricingFormattedInputSchema = z.object({
+  productId: z.string(),
+  userId: z.string().optional(),
+  quantity: z.number(),
+  couponId: z.string().optional(),
+  merchantCoupon: merchantCouponSchema.optional(),
+  upgradeFromPurchaseId: z.string().optional(),
+  code: z.string().optional(),
+  purchases: PurchaseSchema.array(),
+})
+
+const checkForAnyAvailableUpgrades = async ({
+  upgradeFromPurchaseId,
+  productId,
+  purchases,
+}: {
+  upgradeFromPurchaseId: string | undefined
+  productId: string
+  purchases: Purchase[]
+}) => {
+  if (upgradeFromPurchaseId) return upgradeFromPurchaseId
+
+  const {availableUpgradesForProduct} = getSdk()
+
+  const availableUpgrades = await availableUpgradesForProduct(
+    purchases,
+    productId,
+  )
+
+  return find(purchases, (purchase) => {
+    const upgradeProductIds = availableUpgrades.map(
+      (upgrade) => upgrade.upgradableFrom.id,
+    )
+    return upgradeProductIds.includes(purchase.productId)
+  })?.id
+}
```

Comment:
these methods likely belong in the `commerce-server` package

that could be in a follow on PR for sure, but that's where I see them